test/fixed_datetime_formatter_test.dart

```diff
@@ -0,0 +1,104 @@
+  test('Parse year, decade and century', () {
+    var time = FixedDateTimeFormatter('CCEY').decode('1996');
+    expect(time, DateTime.utc(1996));
+  });
+  test('Parse year, century, month', () {
+    var time = FixedDateTimeFormatter('CCYY MM').decode('1996 04');
+    expect(time, DateTime.utc(1996, 4));
+  });
+  test('Parse year, century, month, day', () {
+    var time = FixedDateTimeFormatter('CCYY MM-DD').decode('1996 04-25');
+    expect(time, DateTime.utc(1996, 4, 25));
+  });
+  test('Parse year, century, month, day, hour, minute, second', () {
+    var time = FixedDateTimeFormatter('CCYY MM-DD hh:mm:ss')
+        .decode('1996 04-25 05:03:22');
+    expect(time, DateTime.utc(1996, 4, 25, 5, 3, 22));
+  });
+  test('Parse YYYYMMDDhhmmssSSS', () {
+    var time =
+        FixedDateTimeFormatter('YYYYMMDDhhmmssSSS').decode('19960425050322533');
+    expect(time, DateTime.utc(1996, 4, 25, 5, 3, 22, 533));
+  });
+  test('Parse S', () {
+    var time = FixedDateTimeFormatter('S').decode('1');
+    expect(time, DateTime.utc(0, 1, 1, 0, 0, 0, 100, 0));
+  });
+  test('Parse SS', () {
+    var time = FixedDateTimeFormatter('SS').decode('01');
+    expect(time, DateTime.utc(0, 1, 1, 0, 0, 0, 10, 0));
+  });
+  test('Parse SSS', () {
+    var time = FixedDateTimeFormatter('SSS').decode('001');
+    expect(time, DateTime.utc(0, 1, 1, 0, 0, 0, 1, 0));
+  });
+  test('Parse SSSSSS', () {
+    var time = FixedDateTimeFormatter('SSSSSS').decode('000001');
+    expect(time, DateTime.utc(0, 1, 1, 0, 0, 0, 0, 1));
+  });
+  test('Parse SSSSSS 2', () {
+    var time = FixedDateTimeFormatter('SSSSSS').decode('001000');
+    expect(time, DateTime.utc(0, 1, 1, 0, 0, 0, 1, 0));
+  });
+  test('Parse hex year throws', () {
+    expect(
+      () => FixedDateTimeFormatter('YYYY').decode('0xAB'),
+      throwsFormatException,
+    );
+  });
+  //tryDecode
+  test('Try parse year', () {
+    var time = FixedDateTimeFormatter('YYYY').tryDecode('1996');
+    expect(time, DateTime.utc(1996));
+  });
+  test('Try parse hex year return default', () {
+    var time = FixedDateTimeFormatter('YYYY').tryDecode('0xAB');
+    expect(time, DateTime.utc(0));
+  });
+  test('Try parse invalid returns default', () {
+    var time = FixedDateTimeFormatter('YYYY').tryDecode('1x96');
+    expect(time, DateTime.utc(0));
+  });
+  //encode
+  test('Format simple', () {
+    var time = DateTime.utc(1996, 1);
+    expect('1996 kiwi 01', FixedDateTimeFormatter('YYYY kiwi MM').encode(time));
+  });
+  test('Format YYYYMMDDhhmmss', () {
+    var str = FixedDateTimeFormatter('YYYYMMDDhhmmss')
+        .encode(DateTime.utc(1996, 4, 25, 5, 3, 22));
+    expect('19960425050322', str);
+  });
+  test('Format CCEY-MM', () {
+    var str = FixedDateTimeFormatter('CCEY-MM').encode(DateTime.utc(1996, 4));
+    expect('1996-04', str);
+  });
+  test('Format XCCEY-MMX', () {
+    var str = FixedDateTimeFormatter('XCCEY-MMX').encode(DateTime.utc(1996, 4));
+    expect('X1996-04X', str);
+  });
```

Comment:
Add test of fractional seconds.
Definitely include some edge cases, like:
* `ssS`, `ssSS`, `ssSSS`, `ssSSSS`, `ssSSSSS`, `ssSSSSSS`, `ssSSSSSSS`, with inputs with any number of leading zeros.

Check with negative years. (I guess we can't  parse negative years at all. Which makes some things simpler ... and parsing negative years impossible.)
We can still get negative years as input to `encode`, so we need to make sure it works.
Which likely means removing the sign, because we don't want to start left-padding `"-25"` with zeros. All we can write is digits, so the sign should be ignored.

(Do we need a control character for the sign of the year? `s`/`S` for "sign" are taken, so are "m"/"M" for minus, and `E` for epoch.
Also, what should it write when the year is positive, nothing or space?)

Reply:
I think the concept of negative years for a fixed format parser introduces a whole lot of difficulties as you described, not outweighed by the benefits. I would suggest omitting them completely, throwing an exception when trying to `encode`. `decode` already throws an exception as it checks for the string to be parsed only containing digits.

Reply:
Agree. Just throw on negative year inputs.